Copy

stdavis opened the pull request:
Should probably wait until #38 is merged and let me rebase before merging this...

Closes #39

steveoh commented:
I rebased this for you. I'd like to see the `copy_data` method get invoked before merging this.

stdavis commented:
OK, @steveoh. I've implemented the copy_data call from cli and added some error catching. I ended up refactoring the get_report calls to allow the copy_data to report errors. This is ready for review.

steveoh commented on src/forklift/cli.py:
```diff
@@ -139,11 +139,14 @@ def start_lift(file_path=None):
     lift.process_crates_for(pallets, core.update)
 
+    lift.process_pallets(pallets)
+
```
it might be nice to see the elapsed time of processing and data copying. What do you think?

stdavis replied:
Are you thinking showing the total elapsed time after each of the process and copy calls? What about showing the total time that each of them takes? Something like:
```
process_crates time: 10 min
process_pallets time: 5 min
copy_data time: 7min
```

or something else?

steveoh replied:
I like what you have in your example yes.

steveoh commented on src/forklift/lift.py:
```diff
@@ -77,17 +78,41 @@ def process_pallets(pallets):
+def copy_data(pallets, copy_destinations):
+    '''pallets: Pallets[]
+
+    Loop over all of the pallets and extra the distinct copy_data workspaces.
```
what does this mean?

stdavis replied:
Fixed with 4b5cb16

steveoh commented on tests/test_lift.py:
```diff
@@ -103,3 +95,63 @@ def test_process_pallets_handles_ship_exception(self):
+
+        lift.copy_data([CopyPalletOne(), CopyPalletTwo(), palletThree], ['dest1', 'dest2'])
+
+        self.assertEqual(copytree_mock.call_count, 6)
```
should these counts be 3? CopyPallet 1 and 2 have one.gdb in common and two and three are unique. CopyPallet3 isn't ready to ship so it should not have copy called?

So copytree_mock would get called once for one.gdb and again for two.gdb and three.gdb?

stdavis replied:
There are two destinations so it copy's twice for each copy_data workspace.

steveoh commented on src/forklift/lift.py:
```diff
@@ -77,17 +78,41 @@ def process_pallets(pallets):
+                log.info('copy successful in %s', seat.format_time(clock() - start_seconds))
+            except Exception as e:
+                pallet.success = (False, str(e))
+                log.error('there was an error copying {} to {}: {}'.format(source, destination_workspace, e))
```
with logging it is recommended to use the printf format instead of format. It's also a good idea to add the [exec info](https://github.com/agrc/forklift/blob/ebf9cdec2ad4d5add933dadd9c8c3b545a74bfee/src/forklift/lift.py#L67).

stdavis replied:
Addressed with e9a4c6b.

steveoh commented on src/forklift/lift.py:
```diff
@@ -77,17 +78,41 @@ def process_pallets(pallets):
+                shutil.copytree(source, destination_workspace)
+                log.info('copy successful in %s', seat.format_time(clock() - start_seconds))
+            except Exception as e:
+                pallet.success = (False, str(e))
```
this is going to give us an ugly message. Thoughts on something more readable like 'problem copying or removing data' or using the e.message property?

stdavis replied:
The `e.message` property gave me an empty string in my testing. Not super helpful in trying to diagnose the problem. `str(e)` gave me something pretty nice like: ` [Error 3] The system cannot find the path specified: u'C:\\Temp\\copytest\\scratch.gdb\\*.*'`

steveoh commented:
LGTM
